examples/policy_gradient/reinforce/agent.py

```diff
@@ -2,48 +2,185 @@
 from __future__ import division
 from __future__ import print_function
 
+import gym.spaces
 import tensorflow as tf
 import os
 
+from tensorflow.python.ops.data_flow_ops import StagingArea
+
 import ray
 
+from reinforce.distributions import Categorical, DiagGaussian
 from reinforce.env import BatchedEnv
 from reinforce.policy import ProximalPolicyLoss
 from reinforce.filter import MeanStdFilter
 from reinforce.rollout import rollouts, add_advantage_values
+from reinforce.utils import make_divisible_by
+
+
+def average_gradients(tower_grads):
+  """Calculate the average gradient for each shared variable across all towers.
+  Note that this function provides a synchronization point across all towers.
+  Args:
+    tower_grads: List of lists of (gradient, variable) tuples. The outer list
+      is over individual gradients. The inner list is over the gradient
+      calculation for each tower.
+  Returns:
+     List of pairs of (gradient, variable) where the gradient has been averaged
+     across all towers.
```

Comment:
consider mentioning as a TODO here that we can potentially use NCCL (see https://www.tensorflow.org/performance/performance_models) in the future to speed things up here if it becomes a bottleneck.